### PR TITLE
Fixing bug - netbox_device

### LINKED ISSUE #57397: Adding devices while having multiple racks in Netbox
##### SUMMARY

It appears that you cannot add devices while having more than one rack in Netbox, as ansible  seems to not be able to distinguish between racks.
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

netbox_device
Also other netbox modules can be affected
##### ANSIBLE VERSION

```paste below
ansible 2.8.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15rc1 (default, Nov 12 2018, 14:31:15) [GCC 7.3.0]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Ubuntu 18.04 with Netbox installation v.2.5.12

##### STEPS TO REPRODUCE

Just write short playbook that have to add/delete/modify netbox device while having more that one rack added in Netbox

```yaml
- name: "Test Netbox modules"
  connection: local
  hosts: localhost
  gather_facts: False

 tasks:
    - name: Create R1
       netbox_device:
                 netbox_url: http://netbox.local
                 netbox_token: MyToken
                 data:
                     name: R1
                     rack: rack1
                     position: 12
                     face: front
                     device_type: Force10
                     device_role: Switch1G
                     site: Test_site
                     status: Active
                  state: present
```

##### EXPECTED RESULTS

Add a Netbox device, in this case Force10 1G switch in rack1 at position 12
##### ACTUAL RESULTS

Got error and message as belowe.

```paste below
TASK [netbox : Create R1] ****************************************************************************************************************************************************************************
task path: /root/ansible-provisioning/roles/netbox/tasks/R1.yml:1
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297 `" && echo ansible-tmp-1559727456.73-119493480192297="` echo /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297 `" ) && sleep 0'
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/net_tools/netbox/netbox_device.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-16333oO4kyG/tmpBrcxwu TO /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297/AnsiballZ_netbox_device.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297/ /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297/AnsiballZ_netbox_device.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297/AnsiballZ_netbox_device.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1559727456.73-119493480192297/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_netbox_device_payload_2d7Q1S/__main__.py", line 243, in main
    result = ensure_device_present(nb, nb_endpoint, norm_data)
  File "/tmp/ansible_netbox_device_payload_2d7Q1S/__main__.py", line 267, in ensure_device_present
    data = _find_ids(nb, normalized_data)
  File "/tmp/ansible_netbox_device_payload_2d7Q1S/__main__.py", line 258, in _find_ids
    return find_ids(nb, data)
  File "/tmp/ansible_netbox_device_payload_2d7Q1S/ansible_netbox_device_payload.zip/ansible/module_utils/net_tools/netbox/netbox_utils.py", line 316, in find_ids
    "Multiple results found while searching for key: %s" % (k)

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": {
                "device_role": 1, 
                "device_type": 1, 
                "face": 0, 
                "name": "R1", 
                "position": 12, 
                "rack": "1", 
                "site": 3, 
                "status": 1
            }, : 3, 
                "status": 1
            }, 
            "netbox_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "netbox_url": "http://37.233.99.15", 
            "state": "present", 
            "validate_certs": true
        }
    }, 
    "msg": "Multiple results found while searching for key: rack"
}
            "netbox_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "netbox_url": "http://37.233.99.15", 
            "state": "present", 
            "validate_certs": true
        }
    }, 
    "msg": "Multiple results found while searching for key: rack"
}

PLAY RECAP *******************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

### PR DESCRIPTION
**Pull request by DoctorKM**
##### SUMMARY

Fixing bug that made impossible to add device to Netbox via ansible when Netbox has more than one rack. The netbox_device was attempting to add device in rack using rack's slug, not the name. As racks don't have slug, Netbox was treating that as any rack.
Fixes: #57397 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

netbox_device

### CODE REVIEW

File: test/units/module_utils/net_tools/test_netbox.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 def test_normalize_data():
     assert "name" not in QUERY_TYPES
-    assert QUERY_TYPES.get("rack") == "slug"
+    assert QUERY_TYPES.get("rack") == "name"
```
Comment (FragmentedPacket) [inline test/units/module_utils/net_tools/test_netbox.py:18]: You'll have to change L27 to match the `name` of the rack. ex. `Some rack` so the test passes for it.

**Review by FragmentedPacket — CHANGES_REQUESTED** [review]

**Review by FragmentedPacket — APPROVED** [review]

**Comment by FragmentedPacket** [comment]
!component_name=lib/ansible/module_utils/net_tools/netbox/netbox_utils.py

**Comment by FragmentedPacket** [comment]
shipit

**Comment by FragmentedPacket** [comment]
@s-hertel Here is another one to be merged and fix some issues.

**Comment by minitriga** [comment]
Any update on when this will be merged?

**Comment by FragmentedPacket** [comment]
I would've hoped a few minor versions ago.

I'm in the process of moving everything to Ansible Collections if you want to use it from there and should have the bug fixes that have been merged and haven't been merged. This will be the preferred way of using the Netbox modules moving forward from 2.9 on.

https://galaxy.ansible.com/fragmentedpacket/netbox_modules

Unfortunately it only works with running Ansible from source at the moment

**Comment by richardjohnallsopp** [comment]
Hey can I get an update on where this will be live? I'm running ansible 2.9.1 and still seeing the bug. Or is there a workaround I can implement?

**Comment by FragmentedPacket** [comment]
@richardjohnallsopp I would suggest using the Netbox Modules Ansible Collection from now on. It's under active development.

**Comment by richardjohnallsopp** [comment]
@FragmentedPacket, thanks man I took a look, is there anymore documentation on getting it up and running and on using the new modules? I see you've added loads so I'm really keen to get it working.

**Comment by FragmentedPacket** [comment]
@richardjohnallsopp I do need to update the README.md, but it does have an example of a playbook.

Since you're on Ansible 2.9 you would install the collection via galaxy:
`ansible-galaxy collection install fragmentedpacket.netbox_modules`

And then look at the following doc on how to use them:
https://docs.ansible.com/ansible/latest/user_guide/collections_using.html#id7

You can always open an issue on the github repo for them and we can discuss further there.

**Comment by richardjohnallsopp** [comment]
Thanks man I've opened an issue as I have installed the collections and still see the rack key error.

**Comment by Akasurde** [comment]
Fixed via https://github.com/netbox-community/ansible_modules/commit/becfd7b878a65d821e8e15c7a0c9e0a520ed7496

**Pull request closed without merging**